Fix #135. Add Explicit options for adding ACF Field Groups to the Schema

rsm0128 opened the pull request:
This PR focuses on fixing the experience of mapping ACF Field Groups to the WPGraphQL Schema. 

This video walks through the update: 

https://youtu.be/VvrZGrcwv0Y

---

closes #135

rsm0128 committed:
Fix #135. Add Explicit options for adding ACF Field Groups to the Schema

rsm0128 committed:
Fixed formatting issues. Reverted phpcs auto formatting error and followed WP convention for new code

rsm0128 committed:
fixed spelling issue. Thanks @drewbaker

rsm0128 committed:
updated unit test with new graphql_types_on field

rsm0128 committed:
Aded Explicit Options WpUnit test case

rsm0128 committed:
updated plugin version check

rsm0128 commented:
@jasonbahl 
I added a unit test for the explicit options as we discussed

supriome commented:
When I active this plugin which download from your github repo, I got this warning? Is there some problem? @rsm0128 

`Warning: require(/Users/suprio/react/gatsby-wordpress/src/wp-content/plugins/wp-graphql-acf-bugfix-135/vendor/composer/platform_check.php): failed to open stream: No such file or directory in /Users/suprio/react/gatsby-wordpress/src/wp-content/plugins/wp-graphql-acf-bugfix-135/vendor/composer/autoload_real.php on line 25`

rsm0128 commented:
@supriome You need to run `composer install`.

jasonbahl commented:
@rsm0128 For the "GraphQL Types to show the field group on" option, I think we also need to consider interfaces. 

Specifically the `ContentNode` and `TermNode` interfaces.

For example, if I want a field group, maybe something like SEO Fields, to be available on _all_ Post Types, I could apply the field group to a the ContentNode interface and would be able to query like so: 

```graphql
{
  contentNodes {
    nodes {
      __typename
      id
      acfFieldGroup {
        textField
        textAreaField
      }
    }
  }
}
```

If I were to chose an Interface, possibly the Types that implement that interface would become checked and disabled in the checkboxes: 

![Screen Shot 2021-03-19 at 11 23 34 AM](https://user-images.githubusercontent.com/1260765/111819149-a25e4300-88a5-11eb-881b-c259aef7c45c.png)

jasonbahl commented on src/class-config.php:
```diff
@@ -1188,613 +1114,568 @@ protected function add_field_group_fields( $field_group, $type_name, $layout = f
-					'taxonomy' => $taxonomy,
-				]
-			);
+			$key_prefix = 'post_type__';
```
Is there a specific reason you prefixed these values? 

I think it makes more sense to used the Type name from the schema directly. 

```php
[
  'graphql_types_on' => [
    'Post',
    'Page',
    'Tag',
    'Category',
    'User',
    'Comment',
  ],
]
```

etc

rsm0128 replied:
Users can create custom post type or custom taxonomy with the reserved names. For example, users can't use "User" custom post type if we don't use prefix.

jasonbahl replied:
Types are unique in the GraphQL Schema. You cannot create a `User` Post Type without giving it a unique GraphQL name. 

For example, if I add the following code to register a post type with the `user` GraphQL type, it won't work and there will be a debug message output:

```php
add_action( 'init', function() {

	register_post_type( 'user', [
		'show_in_graphql' => true,
		'graphql_single_name' => 'user',
		'graphql_plural_name' => 'users',
	] );

} );
```

![Screen Shot 2021-03-20 at 8 05 58 AM](https://user-images.githubusercontent.com/1260765/111872112-4520cb00-8953-11eb-8fd5-475bb0594ee7.png)

rsm0128 replied:
OK. I'll remove the prefixes. Thanks for your detailed explanation!

supriome replied:
Any progress? Can this fix be merged into the master branch yet?

rsm0128 replied:
@jasonbahl I found that we need the prefixes to group the post types and taxonomy types on front-end interface. If we don't use the prefixes, it's complicated to manage.

jasonbahl replied:
Should be able to group by Interface, I imagine? 🤔

I'll think on it a bit too

jasonbahl replied:
Also grouping in the UI doesn't have to be defined by the values passed to the field group definition. 

For registering ACF field groups via Code, it would be cleaner for to pass the GraphQL Types directly without having to know about extra prefixes.

rsm0128 replied:
I used `input[value^="post_type__"]` and `input[value^="taxonomy__"]` selector to select all the post types and taxonomies, but if we don't use prefix, it's a bit hard to select.

jasonbahl replied:
@rsm0128 👋 

I pushed a [commit](https://github.com/wp-graphql/wp-graphql-acf/pull/229/commits/f51d294d001e6a4a1846a5f5c67bda362d0c853d) with some re-factoring to the checkboxes. 

This removes the need for prefixing with `post_type__` and `taxonomy__` and uses the GraphQL Type names directly. 

I adjusted the markup to be re-usable for many interfaces, and the jQuery to also be re-usable. 

Right now we have the 2 interfaces (`ContentNode` and `TermNode`), but we'll have _at least_ one more, the `ContentTemplate` Interface, so that ACF field groups can be assigned to Templates in the Schema as well.

jasonbahl replied:
@rsm0128 I pushed [one more commit](https://github.com/wp-graphql/wp-graphql-acf/pull/229/commits/0d2e19126103c72f6f39d0ef1fdb178131d18198) with the `ContentTemplate` interface added. 

Now it works like so (you can see the ContentTemplate interface included now): 

![acf-location-checkboxes](https://user-images.githubusercontent.com/1260765/113090848-e73c8080-91a7-11eb-9579-92c6c8f11955.gif)

jasonbahl commented on tests/wpunit/PostObjectFieldsTest.php:
```diff
@@ -54,6 +54,7 @@ public function register_acf_field_group( $config = [] ) {
 			'description'           => '',
 			'show_in_graphql'       => 1,
 			'graphql_field_name'    => 'postFields',
+			'graphql_types_on'		=> ['post_type__post']
```
Even with these changes, tests are still failing 🤔

rsm0128 replied:
There was a typo and it caused a warning with `user_register_field_groups`. Fixed the warning with new commit.

jasonbahl replied:
Ok, checking it out 👀

jasonbahl replied:
Tests are all still failing, but it's saying that it's because of `undefined function graphql()` so this might be a test environment issue. It seems like WPGraphQL isn't being loaded properly. 

I'll check it out and see if I can get that fixed.

jasonbahl replied:
Ok, so I've confirmed that the tests run locally for me. And all but one pass now. 
So this seems like some sort of CI environment issue.

rsm0128 committed:
fixed typo issue with user register fields

rsm0128 commented:
And I think `\vendor\composer\platform_check.php` should be git ignored or be added to the repo. If I run `composer install` the file marked as an untracked file.

rsm0128 committed:
Added contentNode and termNode support for the explicit option

rsm0128 commented:
@jasonbahl I added ContentNode and TermNode support for explicit options.
Instead of disabling Post Type Checkboxes when ContentNode Checkbox is checked, I implemented it in the same way with Check All button. Please check below screenshot.
![explicit-options](https://user-images.githubusercontent.com/45686588/111863729-a6bb5680-89a0-11eb-81ea-73cba97a6f24.gif)

jasonbahl commented:
@rsm0128 the screenshot for how you implemented the Interface checkbox looks nice! I like it better then the disabled idea I suggested. Good work!

jasonbahl committed:
- add plaftform_check to .gitignore
- update Docker entrypoint to download wp-graphql from WordPress.org
- update test for custom fields

drewbaker commented:
@jasonbahl lets get this bad boy merged! Such a big upgrade to the plugin.

supriome commented:
Hi @rsm0128 , I use your PR to create my `Gatsby` site, But now i found a problem when I add GraphQL Field Name in my posts with some category name, I found it doesn't matter when you query you acf name in `GraphiQL IDE`. Here is my example, you can see whatever i change fields name `filmHiaction` `filmGjyz`, It will return the same result.

![image](https://user-images.githubusercontent.com/19944032/112877611-91a99a80-90f9-11eb-8b3b-48f9ad8bee0a.png)
![image](https://user-images.githubusercontent.com/19944032/112877654-9c642f80-90f9-11eb-8080-194637bd4e83.png)

You can test it with VSCode REST Client extensions with this code
```
POST https://hiiff-web-v4.v-dk.com/graphql
Content-Type: application/json
Authorization: Bearer xxx
X-REQUEST-TYPE: GraphQL

query MyQuery {
  posts(where: {categoryName: "film3"}, first: 1) {
    edges {
      node {
        filmHiaction {
          info {
            color
            category
            name
            time
          }
          fieldGroupName
          description
        }
      }
    }
  }
}
```

Or 
```
POST https://hiiff-web-v4.v-dk.com/graphql
Content-Type: application/json
Authorization: Bearer xxx
X-REQUEST-TYPE: GraphQL

query MyQuery {
  posts(where: {categoryName: "film3"}, first: 1) {
    edges {
      node {
        filmGjyz {
          info {
            color
            category
            name
            time
          }
          fieldGroupName
          description
        }
      }
    }
  }
}
```

You can see the results is no change. Only fields name change.

More detail you can see in this issue i just open. #https://github.com/gatsbyjs/gatsby/issues/30544

jasonbahl committed:
- Update GraphQL Settings fields to use the GraphQL Type names as values instead of prefixing with `post_type__`, `taxonomy__`, etc
- Update to use `graphql_types` instead of `graphql_types_on` setting name
- Update Checkbox markup to be re-usable for multiple interfaces and their possible types
- Update jQuery for checkbox state management to be re-usable for any Interface checkbox set

jasonbahl committed:
- Add the `ContentTemplate` interface to the Checkboxes

jasonbahl committed:
- Fix typo in JS for unsetting checked state

jasonbahl commented:
> @jasonbahl lets get this bad boy merged! Such a big upgrade to the plugin.

@rsm0128 @drewbaker we're getting close for sure!

I pushed some commits refactoring some things with the checkboxes. I'll follow-up with more tomorrow. (I broke the tests 🤦‍♂️ , so I'll fix them) We're getting close to a spot where I'll be comfortable supporting this! 

Stoked to get this out!

jasonbahl committed:
- add filter to `ContentTemplate` fields to pass the source node along as context
- Add ACF Options Pages to the WPGraphQL when the Schema is loading
- Update code to use `graphql_types` instead of `graphql_types_on`
- Update tests to use `graphql_types` instead of `graphql_types_on`
- run `composer install --no-dev`

jasonbahl commented on src/class-config.php:
```diff
@@ -1188,613 +1200,423 @@ protected function add_field_group_fields( $field_group, $type_name, $layout = f
+	 * @since 0.4.1
 	 */
-	protected function add_acf_fields_to_menu_items() {
+	public static function auto_update_field_groups() {
```
This function attempts to update the database, but not all ACF field groups are saved in the database. 

Some ACF Field Groups are JSON files and some are PHP registered.

Instead of running a one-time script to update the database, we should instead check if the `graphql_types` field is set for the type, and if not, we should fallback to the old rules. 

I'm working on a PR to adjust this and account for field groups registered by JSON or PHP, in addition to field groups registered with the UI in the database.

jasonbahl committed:
- removes activation script that tries to upgrade location rules
- Update default `required` value for the `graphql_field_name` rule
- Updates docblocks in class-config
- Updates class-config to get location rules from the location rules class
- Adds new LegacyLocationRules class to help support legacy location rules (before GraphQL Schema Types were explicitly set for each field group)
- Adds JS to handle required field logic for graphql_field_name
- Adds tests for CommentLocation, LegacyRules, MenuItemLocation, MenuLocation, TemplateLocation and TermLocation
- updates composer autoloader map

jasonbahl committed:
- Update test to include nav menu registration

jasonbahl committed:
- Update test message linking to WPGraphQL Issue

jasonbahl committed:
- Update ACF Settings tp include an AJAX action/callback for determining field group location rules
- Move GraphQL Settings into their own meta box
- Register an `AcfFieldGroup` Interface
- Revamp location-rules. They will now be treated as more of a first-class citizen than a legacy thing. Field Groups will still use ACF Location Rules as the first attempt to map field groups to the Schema, but still allow users to override if they choose to.
- Work on JS for conditional visibility of GraphQL Settings fields
- Update LocationRulesTest (instead of LegacyLocationRulesTest)

jasonbahl committed:
Merge commit '3bf411f270c591a89dce3b41acb655046139a10e' into bugfix-135

# Conflicts:
#	docker/app.entrypoint.sh

jasonbahl committed:
- Move GraphQL Field Group Settings into their own box in the ACF Field Group page
- Add Ajax callback for determining rules
- Update Location Rules to check for rule conflicts, and support many edge cases, such as assigning a Field Group to a post_status
- Update JS helpers for the GraphQL Field Group settings. Add conditional logic to the field that allows users to manually chose GraphQL fields or map based on location rules
- Reset location rules when manual graphql type mapping is unchecked.

jasonbahl committed:
- remove dummy code from wp-graphql-acf
- add support for new location rules to map to options pages
- check to see if field group is set to be manually mapped to GraphQL Types, and if so, don't use get_location_rules() to automatically map when building the Schema
- Add test for ACF settings pages and ACF sub settings pages

jasonbahl committed:
- update testing matrix

jasonbahl committed:
- update testing matrix

jasonbahl committed:
- Update tests to test with multiple versions of WPGraphQL

jasonbahl committed:
- Update .gitignore to not ignore the platform check
- back out the change to remove the prefix from field groups (will be addressed later)
